Civi/Api4/Service/Links/MembershipLinksProvider.php

```diff
@@ -0,0 +1,134 @@
+    ];
+  }
+
+  public static function alterMembershipLinksResult(RespondEvent $e): void {
+    $request = $e->getApiRequest();
+    if ($request['version'] == 4 && $request->getEntityName() === 'Membership' && is_a($request, '\Civi\Api4\Action\GetLinks')) {
+      $links = (array) $e->getResponse();
+      $isUpdateBilling = $isCancelSupported = FALSE;
+      $addTemplate = [
+        'ui_action' => '',
+        'entity' => 'Membership',
+        'path' => '',
+        'text' => '',
+        'icon' => 'fa-external-link',
+        'target' => 'crm-popup',
+      ];
+      self::addLinks($links, $addTemplate);
+
+      if (!\CRM_Core_Config::isEnabledBackOfficeCreditCardPayments()) {
+        self::unsetLinks($links, ['followup']);
+      }
+
+      $membershipId = $request->getValue('id');
+      $ownerMembershipId = $request->getValue('owner_membership_id');
+      if ($ownerMembershipId) {
+        self::unsetLinks($links, ['update', 'delete', 'renew', 'followup', 'cancelrecur', 'changebilling']);
+      }
+      elseif ($membershipId) {
+        $paymentObject = \CRM_Financial_BAO_PaymentProcessor::getProcessorForEntity($membershipId, 'membership', 'obj');
+        if (!empty($paymentObject)) {
+          $isUpdateBilling = $paymentObject->supports('updateSubscriptionBillingInfo');
+        }
+        if (!$isUpdateBilling) {
+          self::unsetLinks($links, ['changebilling']);
+        }
+        $isCancelSupported = \CRM_Member_BAO_Membership::isCancelSubscriptionSupported($membershipId);
+        if (!$isCancelSupported) {
+          self::unsetLinks($links, ['cancelrecur']);
+        }
+      }
+
+      // Unset renew and followup for deceased memberships.
+      $membershipStatus = $request->getValue('status_id:name');
+      if ($membershipStatus && $membershipStatus === 'Deceased') {
+        self::unsetLinks($links, ['renew', 'followup']);
+      }
+
+      $membershipTypeId = $request->getValue('membership_type_id');
+      if ($membershipTypeId && \CRM_Financial_BAO_FinancialType::isACLFinancialTypeStatus()) {
+        $finType = \Civi\Api4\MembershipType::get(TRUE)
+          ->addSelect('financial_type_id:name')
+          ->addWhere('id', '=', $membershipTypeId)
+          ->execute()
+          ->first()['financial_type_id:name'] ?? NULL;
+        if ($finType && !\CRM_Core_Permission::check('edit contributions of type ' . $finType)) {
+          self::unsetLinks($links, ['update', 'renew', 'followup']);
+        }
+        if ($finType && !\CRM_Core_Permission::check('delete contributions of type ' . $finType)) {
+          self::unsetLinks($links, ['delete']);
+        }
+      }
+
+      $params = $request->getParams();
+      if (!empty($params['where'][0][0]) && $params['where'][0][0] == 'ui_action') {
+        $action = $params['where'][0][2];
+        $actionLinkIndex = self::getActionIndex($links, $action);
+        $links = [$links[$actionLinkIndex]];
+      }
```

Comment:
@colemanw Since the requirement here is to add new links to the action menu & not the toolbar, the searchkit expects this to return only 1 value and replaces the path link to 0th key [here](https://github.com/civicrm/civicrm-core/blob/master/ext/search_kit/Civi/Api4/Action/SearchDisplay/AbstractRunAction.php#L778). To accommodate this, I had to perform this check and return the value accordingly. Is there a more optimal approach to address this requirement?

Reply:
@jitendrapurohit I think a more optimal approach than this would be to break this into 2 callbacks instead of 1.
Like this:
https://github.com/civicrm/civicrm-core/blob/9316120d32747860b9d9239467b8f2a7c9b22f68/Civi/Api4/Service/Links/ContactLinksProvider.php#L26-L31

The idea is to split the logic between the two callbacks based on the way they each work:

- `civi.api4.getLinks` event gets called first. It's not context-aware, and the results are cached. So in this callback you should unconditionally add all links. This happens *before* the WHERE clause is applied.
- `civi.api.respond` event, happens *after* the WHERE clause is applied, and it is context-aware, so this is where you do the fine-tuning and remove links that are not applicable.

Reply:
ok, pretty sure i started with this approach but was seeing weird number of links in the action menu. Will try this out again & let you know how it goes. Thanks.